### PR TITLE
Big refactoring.

### PR DESCRIPTION
**Pull request by terabytesoftw**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌

### CODE REVIEW

File: src/BaseListView.php [hunk 1]
```diff
@@ -4,491 +4,562 @@
...
+    private string $layoutGridTable = '{items}' . PHP_EOL . '{summary}' . PHP_EOL . '{pager}';
+    private string $pagination = '';
+    protected ?PaginatorInterface $paginator = null;
+    private array $sortLinkAttributes = [];
+    private string $summary = 'gridview.summary';
+    private array $summaryAttributes = [];
+    private TranslatorInterface|null $translator = null;
+    private string $toolbar = '';
+    protected array $urlArguments = [];
+    protected bool $urlEnabledArguments = true;
```
[the remark this answers is not in the document]
Comment (samdark) [reply src/BaseListView.php:36]: Isn't setting `urlArguments` to `[]` enough to disable it?

File: src/BaseListView.php [hunk 2]
```diff
@@ -4,491 +4,562 @@
...
-    public const BULMA = 'bulma';
-    protected array $options = [];
-    protected OffsetPaginator $paginator;
-    protected ?LinkSorter $sorter = null;
-    protected string $summary = 'Showing <b>{begin, number}-{end, number}</b> of <b>{totalCount, number}</b> ' .
-    '{totalCount, plural, one{item} other{items}}';
-    protected array $summaryOptions = [];
-    protected bool $showOnEmpty = false;
+    private array $attributes = [];
+    private bool $container = true;
```
Comment (samdark) [inline src/BaseListView.php:21]:
```suggestion
    private bool $withContainer = true;
```

Comment (samdark) [inline src/BaseListView.php:21]:
```suggestion
    private bool $useContainer = true;
```

File: src/BaseListView.php [hunk 3]
```diff
@@ -4,491 +4,562 @@
...
-     * @param int $currentPage set current page OffsetPaginator::class {@see OffsetPaginator::currentPage()}
+     * Returns a new instance whether container is enabled or not.
+     *
+     * @param bool $value Whether container is enabled or not.
      *
-     * @return $this
+     * @return static
      */
-    public function currentPage(int $currentPage): self
+    public function container(bool $value): static
```
Comment (samdark) [inline src/BaseListView.php:70]: See comments about `$this->container`.

File: src/BaseListView.php [hunk 4]
```diff
@@ -4,491 +4,562 @@
...
-        return $new;
+        return $this->paginator;
     }
 
-    public function getRequestArguments(): ?array
+    public function getTranslator(): TranslatorInterface
     {
-        return $this->requestArguments;
+        if ($this->translator === null) {
+            throw new InvalidArgumentException('The translator is not set.');
```
Comment (samdark) [inline src/BaseListView.php:125]: That's runtime exception, not `InvalidArgumentException`. Also, it is a good place to apply friendly exception.

File: src/BaseListView.php [hunk 5]
```diff
@@ -4,491 +4,562 @@
...
+
+        return $this->translator;
     }
 
-    public function getRequestQueryParams(): ?array
+    public function getUrlGenerator(): UrlGeneratorInterface
     {
-        return $this->requestQueryParams;
+        if ($this->urlGenerator === null) {
+            throw new InvalidArgumentException('Url generator is not set.');
```
Comment (samdark) [inline src/BaseListView.php:134]: That's runtime exception, not `InvalidArgumentException`. Also, it is a good place to apply friendly exception.

File: src/BaseListView.php [hunk 6]
```diff
@@ -4,491 +4,562 @@
...
-     * @throws JsonException
+     * @param bool $value Enabled arguments of the route.
      *
-     * @return string the rendering result
-     *
-     * @see emptyText
+     * @return static
      */
-    protected function renderEmpty(): string
+    public function urlEnabledArguments(bool $value): static
```
Comment (samdark) [inline src/BaseListView.php:370]:
```suggestion
    public function urlEnableArguments(bool $value): static
```

File: src/BaseListView.php [hunk 7]
```diff
@@ -4,491 +4,562 @@
...
-        $options = $this->emptyTextOptions;
+        return $new;
+    }
 
-        /** @psalm-var non-empty-string */
-        $tag = ArrayHelper::remove($options, 'tag', 'div');
+    /**
+     * Return a new instance with url generator interface for pagination.
+     *
+     * @param UrlGeneratorInterface $value The url generator interface for pagination.
```
Comment (samdark) [inline src/BaseListView.php:381]:
```suggestion
     * Return a new instance with URL generator interface for pagination.
     *
     * @param UrlGeneratorInterface $value The URL generator interface for pagination.
```

File: src/BaseListView.php [hunk 8]
```diff
@@ -4,491 +4,562 @@
...
-            case '{pager}':
-                return $this->renderPager();
-            case '{sorter}':
-                return $this->renderSorter();
-            default:
-                return '';
+    protected function run(): string
+    {
+        if (!isset($this->paginator)) {
+            throw new InvalidArgumentException('The "paginator" property must be set.');
```
Comment (samdark) [inline src/BaseListView.php:478]: It is not about invalid argument. It's RuntimeException.

File: src/BaseListView.php [hunk 9]
```diff
@@ -4,491 +4,562 @@
...
+            true => $this->pagination,
+            false => '',
+        };
+    }
 
-        if ($sort === null || empty($sort->getCriteria()) || $this->paginator->getTotalItems() < 1) {
+    private function renderSummary(): string
+    {
+        $pageCount = count($this->getDataReader());
+        /** @var OffsetPaginator */
```
Comment (samdark) [inline src/BaseListView.php:495]: Does it work with KeysetPaginator?

File: src/Column/ActionColumn.php [hunk 10]
```diff
@@ -0,0 +1,458 @@
...
+ * ActionColumn is a column for the {@see GridView} widget that displays buttons for viewing and manipulating the items.
+ */
+final class ActionColumn extends Column
+{
+    private array $buttons = [];
+    private string $primaryKey = 'id';
+    private string $template = '{view}{update}{delete}';
+    private array $urlArguments = [];
+    private Closure|null $urlCreator = null;
+    private bool $urlEnabledArguments = true;
```
Comment (samdark) [inline src/Column/ActionColumn.php:25]:
```suggestion
    private bool $urlEnableArguments = true;
```

Doesn't `$urlArguments = []` disable it already?

File: src/Column/ActionColumn.php [hunk 11]
```diff
@@ -0,0 +1,458 @@
...
+            'view' => static fn (string $url): string => A::tag()
+                ->addAttributes(
+                    [
+                        'name' => 'view',
+                        'role' => 'button',
+                        'style' => 'text-decoration: none!important;',
+                        'title' => 'View',
+                    ],
+                )
+                ->content(Span::tag()->content('&#128270;')->encode(false))
```
Comment (samdark) [inline src/Column/ActionColumn.php:97]: We can use unicode directly, I think i.e. 🔎 It's more readable.

Comment (samdark) [reply src/Column/ActionColumn.php:97]: Then we can remove `->encode(false)`

File: src/Column/ActionColumn.php [hunk 12]
```diff
@@ -0,0 +1,458 @@
...
+    {
+        $label = parent::getLabel();
+
+        return  $label !== '' ? $label : 'Actions';
+    }
+
+    public function getUrlGenerator(): UrlGeneratorInterface
+    {
+        if ($this->urlGenerator === null) {
+            throw new InvalidArgumentException('Url generator is not set.');
```
Comment (samdark) [inline src/Column/ActionColumn.php:145]: RuntimeException?

File: src/Column/CheckboxColumn.php [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+final class CheckboxColumn extends Column
+{
+    private bool $multiple = true;
+
+    /**
+     * Return new instance with the multiple flag is set to false.
+     *
+     * @return self
+     */
+    public function notMultiple(): self
```
Comment (samdark) [inline src/Column/CheckboxColumn.php:22]:
```suggestion
    public function multiple(bool $value): self
```

File: src/Column/Column.php [hunk 14]
```diff
@@ -0,0 +1,389 @@
...
+
+        return $new;
+    }
+
+    /**
+     * Return new instance specifying whether the column is not visible.
+     *
+     * @return static
+     */
+    public function notVisible(): static
```
Comment (samdark) [inline src/Column/Column.php:227]:
```suggestion
    public function visible(bool $value): static
```

File: src/Column/Column.php [hunk 15]
```diff
@@ -0,0 +1,389 @@
...
+         */
+        foreach ($contentAttributes as $name => $value) {
+            if ($value instanceof Closure) {
+                /** @var mixed */
+                $contentAttributes[$name] = $value($data, $key, $index, $this);
+            }
+        }
+
+        if (!array_key_exists('data-label', $contentAttributes) && $this->getLabel() !== '') {
+            $contentAttributes['data-label'] = strtolower($this->getLabel());
```
Comment (samdark) [inline src/Column/Column.php:260]:
```suggestion
            $contentAttributes['data-label'] = mb_strtolower($this->getLabel());
```

File: src/Column/DataColumn.php [hunk 16]
```diff
@@ -0,0 +1,427 @@
...
+
+        return $new;
+    }
+
+    /**
+     * Return new instance whether the column is not sortable.
+     *
+     * @return self
+     */
+    public function notSorting(): self
```
Comment (samdark) [inline src/Column/DataColumn.php:274]:
```suggestion
    public function withSorting(bool $value): self
```

File: src/Column/DataColumn.php [hunk 17]
```diff
@@ -0,0 +1,427 @@
...
+    private string $filter = '';
+    private string $filterAttribute = '';
+    private array $filterInputAttributes = [];
+    /** @psalm-var string[] */
+    private array $filterInputSelectItems = [];
+    private string $filterInputSelectPrompt = '';
+    private string $filterModelName = '';
+    private string $filterType = 'text';
+    /** @psalm-var string[] */
+    private array $filterTypes = [
```
Comment (samdark) [inline src/Column/DataColumn.php:46]: Can I add my own type?

File: src/BaseListView.php [hunk 18]
```diff
@@ -4,491 +4,562 @@
...
-    private TranslatorInterface $translator;
-
-    public function __construct(TranslatorInterface $translator)
-    {
-        $this->translator = $translator;
-    }
+    private array $emptyTextAttributes = [];
+    private string $header = '';
+    private array $headerAttributes = [];
+    private string $layout = '{header}' . PHP_EOL . '{toolbar}';
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/BaseListView.php:26]: done.

File: src/BaseListView.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/BaseListView.php:36]: done.

File: src/BaseListView.php [hunk 19]
```diff
@@ -4,491 +4,562 @@
...
+    private string $toolbar = '';
+    protected array $urlArguments = [];
+    protected bool $urlEnabledArguments = true;
+    protected UrlGeneratorInterface|null $urlGenerator = null;
+    protected string $urlName = '';
+    protected array $urlQueryParameters = [];
 
     /**
-     * Renders the data models.
+     * Renders the data active record classes.
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/BaseListView.php:42]: done.

File: src/BaseListView.php (hunk 2 above)
Comment (terabytesoftw) [reply src/BaseListView.php:21]: done.

Comment (terabytesoftw) [reply src/BaseListView.php:21]: done.

File: src/BaseListView.php (hunk 3 above)
Comment (terabytesoftw) [reply src/BaseListView.php:70]: done.

File: src/BaseListView.php (hunk 4 above)
Comment (terabytesoftw) [reply src/BaseListView.php:125]: done.

File: src/BaseListView.php (hunk 6 above)
Comment (terabytesoftw) [reply src/BaseListView.php:370]: done.

File: src/BaseListView.php (hunk 7 above)
Comment (terabytesoftw) [reply src/BaseListView.php:381]: done.

File: src/BaseListView.php (hunk 9 above)
Comment (terabytesoftw) [reply src/BaseListView.php:495]: In progress.

File: src/Column/ActionColumn.php (hunk 10 above)
Comment (terabytesoftw) [reply src/Column/ActionColumn.php:25]: done.

Comment (terabytesoftw) [reply src/Column/ActionColumn.php:25]: done.

File: src/BaseListView.php (hunk 5 above)
Comment (terabytesoftw) [reply src/BaseListView.php:134]: done.

File: src/Column/ActionColumn.php (hunk 12 above)
Comment (terabytesoftw) [reply src/Column/ActionColumn.php:145]: done.

File: src/Column/ActionColumn.php (hunk 11 above)
Comment (terabytesoftw) [reply src/Column/ActionColumn.php:97]: done.

File: src/Column/CheckboxColumn.php (hunk 13 above)
Comment (terabytesoftw) [reply src/Column/CheckboxColumn.php:22]: done.

File: src/Column/Column.php (hunk 14 above)
Comment (terabytesoftw) [reply src/Column/Column.php:227]: done.

File: src/Column/Column.php (hunk 15 above)
Comment (terabytesoftw) [reply src/Column/Column.php:260]: done.

File: src/Column/DataColumn.php (hunk 16 above)
Comment (terabytesoftw) [reply src/Column/DataColumn.php:274]: done.

File: src/Column/DataColumn.php (hunk 17 above)
Comment (terabytesoftw) [reply src/Column/DataColumn.php:46]: What other types are needed, there are the most used with the input tag, if you want to add a custom filter, you can do it with the `filter()` method.

File: config/web/translator-gridview.php [hunk 20]
```diff
@@ -0,0 +1,16 @@
...
+declare(strict_types=1);
+
+use Yiisoft\Aliases\Aliases;
+use Yiisoft\Translator\CategorySource;
+use Yiisoft\Translator\Message\Php\MessageSource;
+use Yiisoft\Translator\MessageFormatterInterface;
+
+return [
+    'translator.gridview' => static function (Aliases $aliases, MessageFormatterInterface $messageFormatter) {
+        $messageReader = new MessageSource($aliases->get('@gridview/resources/message'));
```
Comment (samdark) [inline config/web/translator-gridview.php:12]: Let's move translations to match what's used in app templates:

```suggestion
        $messageReader = new MessageSource($aliases->get('@gridview/resources/messages'));
```

File: resources/message/en/gridview.php [hunk 21]
```diff
@@ -0,0 +1,7 @@
+<?php
+
+declare(strict_types=1);
+
+return [
+    'gridview.summary' => 'Showing <b>{pageCount}</b> of <b>{totalCount}</b>',
```
Comment (samdark) [inline resources/message/en/gridview.php:6]: How about the following naming?

```suggestion
    'gridview.summary' => 'Page <b>{currentPage}</b> of <b>{totalPages}</b>',
```

File: src/Exception/PaginatorNotSetException.php [hunk 22]
```diff
@@ -0,0 +1,33 @@
...
+        return 'Failed to create widget because "paginator" is not set.';
+    }
+
+    public function getSolution(): ?string
+    {
+        return <<<SOLUTION
+            You can configure the `paginator` property in the widget configuration. You can use `OffSetPaginator::class`
+            or `KeySetPaginator::class` as paginator for more information see the documentation in the link below.
+
+            @link https://github.com/yiisoft/data
```
Comment (samdark) [inline src/Exception/PaginatorNotSetException.php:30]:
```suggestion
            You can configure the `paginator` property in the widget configuration. Use either `OffSetPaginator::class`
            or `KeySetPaginator::class` as paginator. For more information [see the documentation](https://github.com/yiisoft/data).
```

File: src/Exception/TranslatorNotSetException.php [hunk 23]
```diff
@@ -0,0 +1,33 @@
...
+        return 'Failed to create widget because "translator" is not set.';
+    }
+
+    public function getSolution(): ?string
+    {
+        return <<<SOLUTION
+            You can configure the `translator` property in the widget configuration. You can use
+            `TranslatorInterface::class` for more information see the documentation in the link below.
+
+            @link https://github.com/yiisoft/translator
```
Comment (samdark) [inline src/Exception/TranslatorNotSetException.php:30]:
```suggestion
            You can configure the `translator` property in the widget configuration. Use
            `TranslatorInterface::class`. For more information [see the documentation](https://github.com/yiisoft/translator).
```

File: src/Widget/Menu.php [hunk 24]
```diff
@@ -0,0 +1,791 @@
...
+ *
+ * ```php
+ * <?= Menu::Widget()
+ *     ->items([
+ *         ['label' => 'Login', 'link' => 'site/login', 'visible' => true],
+ *     ]);
+ * ?>
+ * ```
+ */
+final class Menu extends Widget
```
Comment (samdark) [inline src/Widget/Menu.php:41]: Doesn't seem to belong to this package.

File: src/Exception/UrlGeneratorNotSetException.php [hunk 25]
```diff
@@ -0,0 +1,33 @@
...
+        return 'Failed to create widget because "urlgenerator" is not set.';
+    }
+
+    public function getSolution(): ?string
+    {
+        return <<<SOLUTION
+            You can configure the `urlgenerator` property in the widget configuration. You can use
+            `UrlGeneratorInterface::class` for more information see the documentation in the link below.
+
+            @link https://github.com/yiisoft/router
```
Comment (samdark) [inline src/Exception/UrlGeneratorNotSetException.php:30]:
```suggestion
            You can configure the `urlgenerator` property in the widget configuration. Use
            `UrlGeneratorInterface::class`. For more information [see the router documentation](https://github.com/yiisoft/router).
```

File: src/Exception/UrlGeneratorNotSetException.php [hunk 26]
```diff
@@ -0,0 +1,33 @@
...
+
+    public function getName(): string
+    {
+        return 'Failed to create widget because "urlgenerator" is not set.';
+    }
+
+    public function getSolution(): ?string
+    {
+        return <<<SOLUTION
+            You can configure the `urlgenerator` property in the widget configuration. You can use
```
Comment (samdark) [inline src/Exception/UrlGeneratorNotSetException.php:27]: Is it `urlGenerator` or `urlgenerator`?

File: src/Widget/Menu.php (hunk 24 above)
Comment (terabytesoftw) [reply src/Widget/Menu.php]: done move to https://github.com/yiisoft/yii-widgets/pull/53.

File: config/web/translator-gridview.php (hunk 20 above)
Comment (terabytesoftw) [reply config/web/translator-gridview.php:12]: done.

File: resources/message/en/gridview.php (hunk 21 above)
Comment (terabytesoftw) [reply resources/message/en/gridview.php:6]: done.

File: src/Exception/UrlGeneratorNotSetException.php (hunk 25 above)
Comment (terabytesoftw) [reply src/Exception/UrlGeneratorNotSetException.php:30]: done.

File: src/Exception/UrlGeneratorNotSetException.php (hunk 26 above)
Comment (terabytesoftw) [reply src/Exception/UrlGeneratorNotSetException.php:27]: `urlGenerator` done.

**Comment by terabytesoftw** [comment]
https://github.com/yiisoft/yii-dataview/pull/92

**Comment by samdark** [comment]
What's left here?

**Review by samdark — APPROVED** [review]

File: src/BaseListView.php [hunk 27]
```diff
@@ -4,491 +4,523 @@
...
+    private string $toolbar = '';
+    protected array $urlArguments = [];
+    protected UrlGeneratorInterface|null $urlGenerator = null;
+    protected string $urlName = '';
+    protected array $urlQueryParameters = [];
+    private bool $withContainer = true;
 
     /**
-     * Renders the data models.
+     * Renders the data active records.
```
Comment (samdark) [inline src/BaseListView.php:44]:
```suggestion
     * Renders the data models.
```

**Comment by samdark** [comment]
🔥

**Pull request merged**